source/develop/release-management/features/storage/incremental-backup.html.md

```diff
@@ -12,173 +12,421 @@ feature_status: Design
-This does not depend on the push or pull modes, but we need qemu
-support for getting the changed blocks from an image.
-
-Basically this is the same as push mode, after qemu pushed the
-changed blocks to external image.
-
-1. oVirt starts local qemu-nbd, mounting the external disk at /dev/nbdX.
-2. oVirt notify the user (e.g. backup application) that the backup
-disk is ready.
-3. Backup application starts a transfer with the relevant disks
-4. Backup application obtains the changed blocks list from
-ovirt-imageio via HTTPS.
-5. ovirt-imageio obtains the changed blocks using qemu-img map and
-the external image.
-6. Backup application downloads the changed blocks (guest blocks)
-from ovirt-imageio via HTTPS.
-7. ovirt-imageio reads the blocks from /dev/nbdX.
+   backup application (not part of oVirt).
 
+1. Backup application create a new disk or a snapshot with existing disk
+   to hold the restored data.
 
-## Links
-- RFC v2 external (pull) backup API:
+1. Backup application starts an upload image transfer for every disk,
+   specifying format=raw. This enable format conversion when uploading
+   raw data to qcow2 disk.
 
-    https://www.redhat.com/archives/libvir-list/2018-April/msg00115.html
+1. Backup application transfer the data included in this restore point
+   to imageio using HTTP API.
 
-- Vmware CBT docs:
+1. Backup application finalize the image transfers.
 
-    https://kb.vmware.com/s/article/1020128
+### Restoring snapshots
 
-- Eric Blake patches:
+Incremental restore will not support restoring snapshots as existed at
+the time of the backup. This limit is common in backup solutions for
+other systems.
 
-    https://www.redhat.com/archives/libvir-list/2018-June/msg01066.html
+### Handling unclean shutdown
+
+If a VM is shut down abnormally, bitmaps on the disk may be left in
+invalid state. Creating incremental backup using such bitmaps will
+lead to corrupt guest data after restore.
+
+To detect bitmap in invalid state we can use the in-use bit in the
+bitmap, but this info is not exposed to the management layer.
+
+To recover from invalid bitmaps, the invalid bitmap and all previous
+bitmaps must be deleted. The next backup will have to include the entire
+disk contents.
+
+## Detailed design
+
+### Incremental backup pipeline
+
+When backing up incremental data, the user always reads raw guest data,
+regardless of the underlying disk format.
+
+qemu exposes the disks using internal NBD server, allowing reading from
+every disk included in the incremental backup:
+
+    http client (raw) <- imageio <- qemu <- image (raw or qcow2)
+
+If the VM is not running, the system will create a paused, stripped-down
+version of the VM, with only backup disks attached, and use libvirt API
+to start and stop the backup.
+
+We considered alternative solution using qemu-nbd, but According to Eric
+Blake qemu-nbd does not support yet exposing bitmap info, so we would
+not be able to provide the change block list.
```

Comment:
Adding a VM in paused mode may be lot of work on both vdsm and engine, so I think the first milestone should not support offline backup.